File: examples/python/Misc/Bootloader/flash_network_bootloader.py
#!/usr/bin/env python3
"""Safely update an RVC2 device's NETWORK bootloader.

By default, this script flashes the NETWORK bootloader bundled with DepthAI to
the recoverable user bootloader slot, leaving the factory bootloader unchanged.
It checks that the connected device is running a compatible NETWORK bootloader,
asks for confirmation, reports flashing progress, and reconnects after the
device reboots to verify that the expected user bootloader is running.

An explicit bootloader image and its expected version can be supplied together.
Passing ``--factory`` instead writes the factory bootloader region, which can
require hardware recovery if interrupted or unsuccessful, so that path requires
additional confirmation and is never enabled by ``--yes`` alone.
"""

import argparse
import sys
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import depthai as dai


CONNECT_ATTEMPTS = 30
CONNECT_RETRY_DELAY_SECONDS = 1.0
VERIFY_ATTEMPTS = 30
VERIFY_RETRY_DELAY_SECONDS = 2.0


class SafetyError(RuntimeError):
    """Raised when a bootloader update fails a safety check."""


@dataclass(frozen=True)
class BootloaderState:
    version: str
    isUserBootloader: bool


def parseArgs() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Safely update an RVC2 device's user NETWORK bootloader using an explicit image "
            "or the image embedded in depthai-core. Factory flashing requires -f and two confirmations."
        ),
    )
    parser.add_argument(
        "--device",
        "-d",
        help="Device IP address, name, or device ID (defaults to the first available device)",
    )
    parser.add_argument(
        "--factory",
        "-f",
        action="store_true",
        help="Flash the factory bootloader instead of the recoverable user bootloader",
    )
    parser.add_argument(
        "--yes",
        "-y",
        action="store_true",
        help="Skip the routine user-slot confirmation (device selection and factory prompts are never skipped)",
    )
    parser.add_argument(
        "--bootloader",
        type=Path,
        help="NETWORK bootloader image to flash instead of the image embedded in depthai",
    )
    parser.add_argument(
        "--expected-version",
        dest="expectedVersion",
        help="Full version expected after flashing --bootloader (for example, 0.0.28+<commit>)",
    )
    args = parser.parse_args()

    if (args.bootloader is None) != (args.expectedVersion is None):
        parser.error("--bootloader and --expected-version must be provided together")
    if args.bootloader is not None and not args.bootloader.is_file():
        parser.error(f"bootloader image does not exist or is not a file: {args.bootloader}")

    return args


def versionString(version: Any) -> str:
    """Compare the full version, including snapshot build metadata."""
    return str(version)


def inspectDevice(bootloader: Any, factory: bool = False) -> BootloaderState:
    if bootloader.getType() != dai.DeviceBootloader.Type.NETWORK:
        raise SafetyError("Refusing to update: the connected device is not running a NETWORK bootloader.")

    if not factory and not bootloader.isUserBootloaderSupported():
        raise SafetyError(
            "Refusing to update: the installed factory bootloader does not support the recoverable "
            "user bootloader slot. Use the M8/USB recovery programming board to update this device; "
            "do not flash its factory bootloader over Ethernet."
        )

    return BootloaderState(
        version=versionString(bootloader.getVersion()),
        isUserBootloader=bootloader.isUserBootloader(),
    )


def deviceLabel(device: Any) -> str:
    if isinstance(device, str):
        return device

    name = getattr(device, "name", "")
    deviceId = getattr(device, "deviceId", "")
    if name and deviceId and name != deviceId:
        return f"{name} ({deviceId})"
    return name or deviceId or str(device)


def selectFirstDevice() -> Any:
    found, device = dai.DeviceBootloader.getFirstAvailableDevice()
    if not found:
        raise RuntimeError("No available DepthAI devices found.")

    print(f"First available device: {deviceLabel(device)}")
    if input("Is this the device you want to flash? [y/N] ").strip().lower() not in {"y", "yes"}:
        return None
    return device


def confirmFactoryAccess(device: Any) -> bool:
    print(f"WARNING: Factory bootloader flashing requested for {deviceLabel(device)}.")
    print("A failed factory bootloader update may require hardware recovery.")
    return input("Allow factory bootloader access? [y/N] ").strip().lower() in {"y", "yes"}


def confirmFlash(device: Any, installed: BootloaderState, targetVersion: str, factory: bool = False) -> bool:
    print(f"Device: {deviceLabel(device)}")
    print(f"Installed NETWORK bootloader: {installed.version} is user bootloader: {installed.isUserBootloader}")
    print(f"Target NETWORK bootloader:    {targetVersion}")
    if factory:
        print("Target: FACTORY bootloader region")
        print("WARNING: This operation can make the device unbootable and require hardware recovery.")
    else:
        print("Target: recoverable USER bootloader slot (factory bootloader will not be modified)")
    print("Do not disconnect the device or interrupt its power while updating.")
    prompt = "Flash the factory bootloader? [y/N] " if factory else "Continue? [y/N] "
    response = input(prompt).strip().lower()
    return response in {"y", "yes"}


def printProgress(progress: float, factory: bool = False) -> None:
    target = "factory" if factory else "user"
    print(f"\rUpdating {target} bootloader: {progress * 100:5.1f}%", end="", flush=True)


def connectBootloader(device: Any, factory: bool = False) -> Any:
    lastError = "device did not become available"

    for attempt in range(CONNECT_ATTEMPTS):
        try:
            return dai.DeviceBootloader(device, factory)
        except RuntimeError as exc:
            if "Specified device not found" not in str(exc):
                raise
            lastError = str(exc)

        if attempt + 1 < CONNECT_ATTEMPTS:
            print(
                f"Device is reconnecting; retrying discovery "
                f"({attempt + 1}/{CONNECT_ATTEMPTS})...",
                file=sys.stderr,
            )
            time.sleep(CONNECT_RETRY_DELAY_SECONDS)

    raise RuntimeError(f"Device did not reconnect after {CONNECT_ATTEMPTS} attempts: {lastError}")


def verifyUpdate(device: Any, targetVersion: str) -> None:
    lastError = "Device did not reconnect"

    for attempt in range(VERIFY_ATTEMPTS):
        if attempt:
            time.sleep(VERIFY_RETRY_DELAY_SECONDS)

        try:
            with dai.DeviceBootloader(device, False) as bootloader:
                state = inspectDevice(bootloader)
        except Exception as exc:
            lastError = str(exc)
            continue

        if state.version != targetVersion:
            lastError = f"Device reported version {state.version}, expected {targetVersion}"
            continue
        if not state.isUserBootloader:
            lastError = "Device rebooted into its factory bootloader instead of the updated user bootloader"
            continue

        return

    raise RuntimeError(f"Update was written but post-reboot verification failed: {lastError}")


def main() -> int:
    args = parseArgs()
    targetVersion = args.expectedVersion or versionString(dai.DeviceBootloader.getEmbeddedBootloaderVersion())
    bootloaderPath = str(args.bootloader) if args.bootloader is not None else ""

    try:
        device = args.device
        if device is None:
            device = selectFirstDevice()
            if device is None:
                print("You can select a specific device using the --device argument.")
                return 0

        if args.factory and not confirmFactoryAccess(device):
            print("Cancelled.")
            return 0

        with connectBootloader(device, args.factory) as bootloader:
            installed = inspectDevice(bootloader, args.factory)

            # if installed.version == targetVersion and installed.isUserBootloader:
            #     print(f"NETWORK bootloader is already up to date ({targetVersion}).")
            #     return 0

            if (args.factory or not args.yes) and not confirmFlash(
                device, installed, targetVersion, args.factory
            ):
                print("Cancelled.")
                return 0

            def progressCallback(progress: float) -> None:
                printProgress(progress, args.factory)

            if args.factory:
                success, error = bootloader.flashBootloader(progressCallback, bootloaderPath)
            else:
                success, error = bootloader.flashUserBootloader(progressCallback, bootloaderPath)
            print()
            if not success:
                target = "factory" if args.factory else "user"
                raise RuntimeError(f"Device rejected or failed to verify the {target} bootloader update: {error}")

        if args.factory:
            print("The device reported that the factory bootloader flash completed successfully.")
        else:
            print("Flash verification succeeded; waiting for the device to reboot and checking the running version...")
            verifyUpdate(device, targetVersion)
    except SafetyError as exc:
        print(str(exc), file=sys.stderr)
        return 2
    except Exception as exc:
        print(f"Failed to update the NETWORK bootloader: {exc}", file=sys.stderr)
        if args.factory:
            print("Factory bootloader flashing was enabled; inspect the device state before retrying.", file=sys.stderr)
        else:
            print("The factory bootloader was not modified by this script.", file=sys.stderr)
        return 1

    if args.factory:
        print(f"NETWORK factory bootloader {targetVersion} was flashed successfully.")
    else:
        print(f"NETWORK user bootloader {targetVersion} is running and verified.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/cpp/Misc/Bootloader/flash_network_bootloader.cpp
#include <algorithm>
#include <argparse/argparse.hpp>
#include <cctype>
#include <chrono>
#include <cstdlib>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <memory>
#include <optional>
#include <stdexcept>
#include <string>
#include <thread>
#include <tuple>
#include <utility>

#include "depthai/depthai.hpp"

namespace {

constexpr int CONNECT_ATTEMPTS = 30;
constexpr auto CONNECT_RETRY_DELAY = std::chrono::seconds(1);
constexpr int VERIFY_ATTEMPTS = 30;
constexpr auto VERIFY_RETRY_DELAY = std::chrono::seconds(2);

class SafetyError : public std::runtime_error {
   public:
    using std::runtime_error::runtime_error;
};

struct BootloaderState {
    std::string version;
    bool isUserBootloader;
};

struct SelectedDevice {
    dai::DeviceInfo info;
    std::string label;
};

struct Options {
    std::string device;
    bool factory = false;
    bool skipConfirmation = false;
    std::filesystem::path bootloaderPath;
    std::optional<std::string> expectedVersion;
};

std::string deviceLabel(const dai::DeviceInfo& device) {
    if(!device.name.empty() && !device.deviceId.empty() && device.name != device.deviceId) {
        return device.name + " (" + device.deviceId + ")";
    }
    if(!device.name.empty()) return device.name;
    if(!device.deviceId.empty()) return device.deviceId;
    return device.toString();
}

bool confirm(const std::string& prompt) {
    std::cout << prompt << std::flush;

    std::string response;
    if(!std::getline(std::cin, response)) return false;

    const auto first = std::find_if_not(response.begin(), response.end(), [](unsigned char character) { return std::isspace(character); });
    const auto last = std::find_if_not(response.rbegin(), response.rend(), [](unsigned char character) { return std::isspace(character); }).base();
    if(first >= last) return false;

    response = std::string(first, last);
    std::transform(response.begin(), response.end(), response.begin(), [](unsigned char character) { return static_cast<char>(std::tolower(character)); });
    return response == "y" || response == "yes";
}

Options parseArgs(int argc, char** argv) {
    argparse::ArgumentParser program("flash_network_bootloader", "1.0.0");
    program.add_description(
        "Safely update an RVC2 device's user NETWORK bootloader using an explicit image or the image embedded in depthai-core. Factory flashing "
        "requires -f and two confirmations.");
    program.add_argument("--device", "-d")
        .default_value(std::string(""))
        .help("Device IP address, name, or device ID (defaults to the first available device)");
    program.add_argument("--factory", "-f").flag().help("Flash the factory bootloader instead of the recoverable user bootloader");
    program.add_argument("--yes", "-y").flag().help("Skip the routine user-slot confirmation (device selection and factory prompts are never skipped)");
    program.add_argument("--bootloader").help("NETWORK bootloader image to flash instead of the image embedded in depthai");
    program.add_argument("--expected-version").help("Full version expected after flashing --bootloader (for example, 0.0.28+<commit>)");

    try {
        program.parse_args(argc, argv);
    } catch(const std::runtime_error& error) {
        std::cerr << error.what() << '\n';
        std::cerr << program;
        throw;
    }

    const bool hasBootloader = program.is_used("--bootloader");
    const bool hasExpectedVersion = program.is_used("--expected-version");
    if(hasBootloader != hasExpectedVersion) {
        std::cerr << "--bootloader and --expected-version must be provided together\n";
        std::cerr << program;
        throw std::invalid_argument("--bootloader and --expected-version must be provided together");
    }

    Options options{
        program.get<std::string>("--device"),
        program.get<bool>("--factory"),
        program.get<bool>("--yes"),
        {},
        std::nullopt,
    };

    if(hasBootloader) {
        options.bootloaderPath = program.get<std::string>("--bootloader");
        options.expectedVersion = program.get<std::string>("--expected-version");
        if(!std::filesystem::is_regular_file(options.bootloaderPath)) {
            std::cerr << "Bootloader image does not exist or is not a file: " << options.bootloaderPath << '\n';
            std::cerr << program;
            throw std::invalid_argument("Bootloader image does not exist or is not a file: " + options.bootloaderPath.string());
        }
    }

    return options;
}

std::optional<SelectedDevice> selectDevice(const std::string& deviceArgument) {
    if(!deviceArgument.empty()) {
        return SelectedDevice{dai::DeviceInfo(deviceArgument), deviceArgument};
    }

    bool found = false;
    dai::DeviceInfo device;
    std::tie(found, device) = dai::DeviceBootloader::getFirstAvailableDevice();
    if(!found) {
        throw std::runtime_error("No available DepthAI devices found.");
    }

    const auto label = deviceLabel(device);
    std::cout << "First available device: " << label << '\n';
    if(!confirm("Is this the device you want to flash? [y/N] ")) return std::nullopt;

    return SelectedDevice{std::move(device), label};
}

bool confirmFactoryAccess(const SelectedDevice& device) {
    std::cout << "WARNING: Factory bootloader flashing requested for " << device.label << ".\n";
    std::cout << "A failed factory bootloader update may require hardware recovery.\n";
    return confirm("Allow factory bootloader access? [y/N] ");
}

BootloaderState inspectDevice(dai::DeviceBootloader& bootloader, bool factory = false) {
    if(bootloader.getType() != dai::DeviceBootloader::Type::NETWORK) {
        throw SafetyError("Refusing to update: the connected device is not running a NETWORK bootloader.");
    }

    if(!factory && !bootloader.isUserBootloaderSupported()) {
        throw SafetyError(
            "Refusing to update: the installed factory bootloader does not support the recoverable user bootloader slot. Use the M8/USB recovery "
            "programming board to update this device; do not flash its factory bootloader over Ethernet.");
    }

    return BootloaderState{bootloader.getVersion().toString(), bootloader.isUserBootloader()};
}

bool confirmFlash(const SelectedDevice& device, const BootloaderState& installed, const std::string& targetVersion, bool factory) {
    std::cout << "Device: " << device.label << '\n';
    std::cout << "Installed NETWORK bootloader: " << installed.version << " is user bootloader: " << (installed.isUserBootloader ? "true" : "false") << '\n';
    std::cout << "Target NETWORK bootloader:    " << targetVersion << '\n';
    if(factory) {
        std::cout << "Target: FACTORY bootloader region\n";
        std::cout << "WARNING: This operation can make the device unbootable and require hardware recovery.\n";
    } else {
        std::cout << "Target: recoverable USER bootloader slot (factory bootloader will not be modified)\n";
    }
    std::cout << "Do not disconnect the device or interrupt its power while updating.\n";
    return confirm(factory ? "Flash the factory bootloader? [y/N] " : "Continue? [y/N] ");
}

void printProgress(float progress, bool factory) {
    std::cout << "\rUpdating " << (factory ? "factory" : "user") << " bootloader: " << std::fixed << std::setw(5) << std::setprecision(1) << progress * 100.0F
              << '%' << std::flush;
}

std::unique_ptr<dai::DeviceBootloader> connectBootloader(const dai::DeviceInfo& device, bool factory) {
    std::string lastError = "device did not become available";

    for(int attempt = 0; attempt < CONNECT_ATTEMPTS; ++attempt) {
        try {
            return std::make_unique<dai::DeviceBootloader>(device, factory);
        } catch(const std::runtime_error& error) {
            if(std::string(error.what()).find("Specified device not found") == std::string::npos) throw;
            lastError = error.what();
        }

        if(attempt + 1 < CONNECT_ATTEMPTS) {
            std::cerr << "Device is reconnecting; retrying discovery (" << attempt + 1 << '/' << CONNECT_ATTEMPTS << ")...\n";
            std::this_thread::sleep_for(CONNECT_RETRY_DELAY);
        }
    }

    throw std::runtime_error("Device did not reconnect after " + std::to_string(CONNECT_ATTEMPTS) + " attempts: " + lastError);
}

void verifyUpdate(const dai::DeviceInfo& device, const std::string& targetVersion) {
    std::string lastError = "Device did not reconnect";

    for(int attempt = 0; attempt < VERIFY_ATTEMPTS; ++attempt) {
        if(attempt != 0) std::this_thread::sleep_for(VERIFY_RETRY_DELAY);

        try {
            dai::DeviceBootloader bootloader(device, false);
            const auto state = inspectDevice(bootloader);
            if(state.version != targetVersion) {
                lastError = "Device reported version " + state.version + ", expected " + targetVersion;
                continue;
            }
            if(!state.isUserBootloader) {
                lastError = "Device rebooted into its factory bootloader instead of the updated user bootloader";
                continue;
            }
            return;
        } catch(const std::exception& error) {
            lastError = error.what();
        }
    }

    throw std::runtime_error("Update was written but post-reboot verification failed: " + lastError);
}

}  // namespace

int main(int argc, char** argv) {
    Options options{};
    try {
        options = parseArgs(argc, argv);
    } catch(const std::exception&) {
        return EXIT_FAILURE;
    }

    const auto targetVersion = options.expectedVersion.value_or(dai::DeviceBootloader::getEmbeddedBootloaderVersion().toString());

    try {
        const auto selectedDevice = selectDevice(options.device);
        if(!selectedDevice) {
            std::cout << "You can select a specific device using the --device argument.\n";
            return EXIT_SUCCESS;
        }

        if(options.factory && !confirmFactoryAccess(*selectedDevice)) {
            std::cout << "Cancelled.\n";
            return EXIT_SUCCESS;
        }

        {
            auto bootloader = connectBootloader(selectedDevice->info, options.factory);
            const auto installed = inspectDevice(*bootloader, options.factory);

            if((options.factory || !options.skipConfirmation) && !confirmFlash(*selectedDevice, installed, targetVersion, options.factory)) {
                std::cout << "Cancelled.\n";
                return EXIT_SUCCESS;
            }

            const auto progressCallback = [&options](float progress) { printProgress(progress, options.factory); };
            bool success = false;
            std::string error;
            if(options.factory) {
                std::tie(success, error) = bootloader->flashBootloader(progressCallback, options.bootloaderPath);
            } else {
                std::tie(success, error) = bootloader->flashUserBootloader(progressCallback, options.bootloaderPath);
            }
            std::cout << '\n';

            if(!success) {
                throw std::runtime_error("Device rejected or failed to verify the " + std::string(options.factory ? "factory" : "user")
                                         + " bootloader update: " + error);
            }
        }

        if(options.factory) {
            std::cout << "The device reported that the factory bootloader flash completed successfully.\n";
        } else {
            std::cout << "Flash verification succeeded; waiting for the device to reboot and checking the running version...\n";
            verifyUpdate(selectedDevice->info, targetVersion);
        }
    } catch(const SafetyError& error) {
        std::cerr << error.what() << '\n';
        return 2;
    } catch(const std::exception& error) {
        std::cerr << "Failed to update the NETWORK bootloader: " << error.what() << '\n';
        if(options.factory) {
            std::cerr << "Factory bootloader flashing was enabled; inspect the device state before retrying.\n";
        } else {
            std::cerr << "The factory bootloader was not modified by this example.\n";
        }
        return EXIT_FAILURE;
    }

    if(options.factory) {
        std::cout << "NETWORK factory bootloader " << targetVersion << " was flashed successfully.\n";
    } else {
        std::cout << "NETWORK user bootloader " << targetVersion << " is running and verified.\n";
    }
    return EXIT_SUCCESS;
}
